### ISSUE TITLE
Pool TLS connections aren't long-lived anymore

### ISSUE THREAD

**Issue by rentziass** [redis/go-redis#3137]
👋 We've been investigating an increase in CPU load both in our apps and our Redis clusters for a while now, and after noticing that a big portion of our CPU time was being spent on TLS handshakes we started looking at our connections and noticed exorbitant amounts of [stale connections](https://github.com/redis/go-redis/blob/d9eeed131a32b4aeaf1e1915b6717f8cd2af9124/internal/pool/pool.go?plain=1#L427). Importantly, these are all **TLS connections** from connections pools. We tracked it down to us upgrading to go-redis 9.5.4 and when we downgraded to 9.5.2 we went back to the usual occasional stale connection:

![Image](https://github.com/user-attachments/assets/c98c35b5-4346-43a3-be93-c60a79603889)

## Expected Behavior

Under normal conditions have only a handful of connections being reported as stale.

## Current Behavior

Connections are barely being re-used.

## Possible Solution

It looks like the issue started with [this commit](https://github.com/redis/go-redis/commit/0858ed24e6d08c1940bcb86ccbf8913c761d9ec0), specifically [reading directly](https://github.com/redis/go-redis/blob/0858ed24e6d08c1940bcb86ccbf8913c761d9ec0/internal/pool/conn_check.go#L37) from the `(*tls.Conn).NetConn()`; from [`NetConn()` docs](https://pkg.go.dev/crypto/tls#Conn.NetConn):

> NetConn returns the underlying connection that is wrapped by c. Note that writing to or reading from this connection directly will corrupt the TLS session.

Unfortunately, until [Go supports non-blocking readable connection testing](https://github.com/golang/go/issues/15735) I'm not sure there's a good reliable way to check the health of TLS connections here. Until then reverting this direct read on the underlying connection is probably the way to go?

MySQL driver has also had a [similar issue](https://github.com/go-sql-driver/mysql/issues/1392).

## Steps to Reproduce

1. Create a ClusterClient, connecting to a cluster using TLS (ours is an Azure Enterprise cluster, not that it should make a difference). Configuring the client to have MinIdleConns adds a multiplier to this and makes much more apparent.
2. Observe the number of stale connections grow linearly over time as more and more operations are done
3. Downgrade to `< 9.5.4` (9.5.2 is the first prior available version right now) and observe almost no stale connection (with the cluster in healthy conditions of course

## Detailed Description

Reading directly from the underlying `net.Conn` of a TLS connection as part of `connCheck` is corrupting the TLS session, resulting in the closing of tons of (eventually all) TLS connections in connection pools that then have to be recreated.

We can fork go-redis and add some more instrumentation to measure the reasons why connections are being closed but I'm fairly positive `unexpected read from socket` is the error `connCheck` is hitting.

**Comment by rentziass** [comment]
Update on this: we deployed a patched version that removes the read from the underlying net.Conn (essentially https://github.com/redis/go-redis/pull/3138) and we confirmed that fixes the issue 🎉

**Comment by vladvildanov** [comment]
@rentziass Thank you for your valuable input! I would make sure we include this bug fix into our next 9.6.2 release

**Comment by rentziass** [comment]
Hey @vladvildanov 👋 Thanks for reviewing and merging the PR! Do we need those changes ported to the [9.6 branch](https://github.com/redis/go-redis/blob/v9.6/internal/pool/conn_check.go)? Code [changed a bit](https://github.com/redis/go-redis/pull/3066/files) in the meantime on the branch, do you need me to open another PR against that branch or do you want to do it?

**Comment by vladvildanov** [comment]
@rentziass Yes, this would be nice

**Comment by rentziass** [comment]
Backported in https://github.com/redis/go-redis/pull/3145, I'm gonna test it real quick to make sure this version is also working as expected.

**Comment by rentziass** [comment]
@vladvildanov just confirmed #3145 works with our cluster as well!

**Comment by rentziass** [comment]
@vladvildanov do you know roughly what the ETA of releasing v9.6.2 would be? Sorry for the spam, we're blocked on upgrading past 9.5.2 right now :/

**Comment by vladvildanov** [comment]
@rentziass I'll try to make it tomorrow

**Comment by rentziass** [comment]
Thank you ❤ and anything I can do to help, please let me know!

**Comment by vladvildanov** [comment]
@rentziass 9.6.2 just got released
